### PR TITLE
[SPARK-13871][SQL] Support for inferring filters from data constraints

### PR DESCRIPTION
**Pull request by sameeragarwal**
## What changes were proposed in this pull request?

This PR generalizes the `NullFiltering` optimizer rule in catalyst to `InferFiltersFromConstraints` that can automatically infer all relevant filters based on an operator's constraints while making sure of 2 things:

(a) no redundant filters are generated, and 
(b) filters that do not contribute to any further optimizations are not generated.
## How was this patch tested?

Extended all tests in `InferFiltersFromConstraintsSuite` (that were initially based on `NullFilteringSuite` to test filter inference in `Filter` and `Join` operators.

In particular the 2 tests ( `single inner join with pre-existing filters: filter out values on either side` and `multiple inner joins: filter out values on all sides on equi-join keys` attempts to highlight/test the real potential of this rule for join optimization.

### CODE REVIEW

**Comment by sameeragarwal** [comment]
cc @nongli @yhuai @gatorsmile

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 1]
```diff
@@ -598,50 +598,44 @@ object NullPropagation extends Rule[LogicalPlan] {
...
-      val newRightChild = if (rightIsNotNullConstraints.nonEmpty) {
-        Filter(rightIsNotNullConstraints.reduce(And), right)
-      } else {
-        right
-      }
-      if (newLeftChild != left || newRightChild != right) {
-        Join(newLeftChild, newRightChild, joinType, condition)
-      } else {
-        join
+    case join@Join(left, right, joinType, conditionOpt) =>
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Nit: `join @ Join`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 2]
```diff
@@ -598,50 +598,44 @@ object NullPropagation extends Rule[LogicalPlan] {
...
   def apply(plan: LogicalPlan): LogicalPlan = plan transform {
     case filter @ Filter(condition, child) =>
-      // We generate a list of additional isNotNull filters from the operator's existing constraints
-      // but remove those that are either already part of the filter condition or are part of the
-      // operator's child constraints.
-      val newIsNotNullConstraints = filter.constraints.filter(_.isInstanceOf[IsNotNull]) --
+      // For the Filter operator, we try to generate additional filters by only inferring the
+      // IsNotNull constraints. These IsNotNull filters are then used while generating the
+      // physical plan to quickly short circuit the null checks in the generated code.
+      val newFilters = filter.constraints.filter(_.isInstanceOf[IsNotNull]) --
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: We need to infer the Filter conditions too.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 3]
```diff
@@ -598,50 +598,44 @@ object NullPropagation extends Rule[LogicalPlan] {
...
+      val additionalConstraints = join.constraints.filter { c =>
+        // Only consider constraints that can be pushed down to either the left or the right child
+        c.references.subsetOf(left.outputSet) || c.references.subsetOf(right.outputSet)} --
+        (left.constraints ++ right.constraints)
+      val newConditionOpt = conditionOpt match {
+        case Some(condition) =>
+          val newFilters = additionalConstraints -- splitConjunctivePredicates(condition)
+          if (newFilters.nonEmpty) Option(And(newFilters.reduce(And), condition)) else None
+        case None =>
+          if (additionalConstraints.nonEmpty) Option(additionalConstraints.reduce(And)) else None
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Nit: `Option` -> `Some`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 4]
```diff
@@ -598,50 +598,44 @@ object NullPropagation extends Rule[LogicalPlan] {
...
-        join
+    case join@Join(left, right, joinType, conditionOpt) =>
+      val additionalConstraints = join.constraints.filter { c =>
+        // Only consider constraints that can be pushed down to either the left or the right child
+        c.references.subsetOf(left.outputSet) || c.references.subsetOf(right.outputSet)} --
+        (left.constraints ++ right.constraints)
+      val newConditionOpt = conditionOpt match {
+        case Some(condition) =>
+          val newFilters = additionalConstraints -- splitConjunctivePredicates(condition)
+          if (newFilters.nonEmpty) Option(And(newFilters.reduce(And), condition)) else None
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Nit: `Option` -> `Some`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 2 above)
Comment (sameeragarwal) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: This was actually intentional as I couldn't think of an optimization that'd benefit from inferring other filter conditions. Do you have something on your mind?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 3 above)
Comment (sameeragarwal) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: We prefer `Option(x)` over `Some(x)` as the former handles null values better (please see https://github.com/databricks/scala-style-guide#options). Here is how scala handles Option(x):

```scala
  /** An Option factory which creates Some(x) if the argument is not null,
   *  and None if it is null.
   *
   *  @param  x the value
   *  @return   Some(value) if value != null, None if value == null
   */
  def apply[A](x: A): Option[A] = if (x == null) None else Some(x)
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 2 above)
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: The rules `PushPredicateThroughJoin` and `OuterJoinElimination` could utilize the inferred conditions in `Filter`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 4 above)
Comment (sameeragarwal) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Same as above, although here let me use `reduceOption` instead :)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 3 above)
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: uh, I see. In the code base, most of codes are using `Some`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 2 above)
Comment (sameeragarwal) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Can you please give an example where non-IsNotNull filter conditions might be useful? Thanks!

Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: For example, given ```case f @ Filter(filterCondition, Join(left, right, joinType, joinCondition))``` When the join type is left/right outer joins, the common filter conditions are unable to push down into Join condition. 

In these cases, we can infer some left/right filter condition by using the common filter conditions, we can push down more conditions into `Join`.

Comment (sameeragarwal) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: I think I see what you mean. Queries of the form: `x.join(y, Inner, Some(x.a === y.a)).where(x.a > 5)` can be optimized to `x.where('a > 5).join(y.where('a > 5), Inner, Some(x.a === y.a)`. Thanks, I'll add it back.

Comment (sameeragarwal) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: By the way, slightly unrelated to this discussion but Yin, Davies and I had an offline discussion earlier today and we are considering generating the `isNotNull` filters in the physical plan to prevent some TPCDS regressions due to filter pushdown.

**Comment by sameeragarwal** [comment]
test this please

**Comment by sameeragarwal** [comment]
test this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 2 above)
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Thank you for letting me know it. Will you do it? Or you want me to do it? Thanks!

Comment (sameeragarwal) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: yes, please see the last commit 4b0cf5f581901919a5513554b2982ef327a4d87a

Comment (sameeragarwal) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: I'll have the physical plan changes in a separate PR (Need to look into TPCDS q56 and q59)

Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Great! Will expect to see your PR. Thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 5]
```diff
@@ -78,7 +78,7 @@ abstract class Optimizer extends RuleExecutor[LogicalPlan] {
       CombineLimits,
       CombineUnions,
       // Constant folding and strength reduction
-      NullFiltering,
+      InferFiltersFromConstraints,
```
Comment (nongli) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: This rule doesn't seem related to the comment above it (same for nullfiltering)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 6]
```diff
@@ -598,50 +598,41 @@ object NullPropagation extends Rule[LogicalPlan] {
...
+ * optimizing the query) by inserting relevant filters in the query plan based on an operator's
+ * data constraints. These filters are currently inserted to the existing conditions in the Filter
+ * operators and on either side of Join operators.
  *
  * Note: While this optimization is applicable to all types of join, it primarily benefits Inner and
  * LeftSemi joins.
  */
-object NullFiltering extends Rule[LogicalPlan] with PredicateHelper {
+object InferFiltersFromConstraints extends Rule[LogicalPlan] with PredicateHelper {
+  // We generate a list of additional filters from the operator's existing constraint but remove
```
Comment (nongli) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: This comment seems like a clearer version of the first sentence in the object comment. I'd remove this here and replace the first sentence of the above comment.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 7]
```diff
@@ -598,50 +598,41 @@ object NullPropagation extends Rule[LogicalPlan] {
...
-        right
-      }
-      if (newLeftChild != left || newRightChild != right) {
-        Join(newLeftChild, newRightChild, joinType, condition)
-      } else {
-        join
+    case join @ Join(left, right, joinType, conditionOpt) =>
+      val additionalConstraints = join.constraints.filter { c =>
+        // Only consider constraints that can be pushed down to either the left or the right child
+        c.references.subsetOf(left.outputSet) || c.references.subsetOf(right.outputSet)} --
```
Comment (nongli) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: I think this line is really hard to read. Split the join.constraints.filter{} line from the set difference line and us another local var

**Comment by sameeragarwal** [comment]
thanks, all comments addressed!

**Comment by nongli** [comment]
LGTM
